### DISCUSSION TITLE
CTEs in MS SQL

### DISCUSSION THREAD

**Discussion by bloukanov** [sfu-db/connector-x#156] [Ideas]
Hi there,

Loving the package so far. I'd like to now see if we can use it for larger queries with intermediary tables. But when I try to do a single `select` with any CTE's (https://docs.microsoft.com/en-us/sql/t-sql/queries/with-common-table-expression-transact-sql?view=sql-server-ver15), it fails with the error:

```RuntimeError: Token error: 'Incorrect syntax near the keyword 'WITH'.' on server X executing  on line 1 (code: 156, state: 1, class: 15)```

The same select works in pandas read_sql. Would it be possible to make this work or no?

**Comment by wangxiaoying** [comment]
Hi @bloukanov , thanks for the suggestion! For now we only support single select query. Adding support for CTE and others like view creation will bring challenge to the query partition part since we need to figure out which query should be partitioned, and we also need to support non-selection queries. We will look into the need and possible solutions of this.

**Reply by bloukanov** [reply]
sounds good!

**Reply by wangxiaoying** [reply]
Hi @bloukanov , I took a deeper look at this error and found that I misunderstood the reason that cause the error earlier. We can support CTE after some modification on our query generation logic. (Here is the PR: https://github.com/sfu-db/connector-x/pull/161).

This will be added to our next release. (Please let me know if you need it earlier, then I can release an alpha version first.) Thanks for bringing up this issue!

**Reply by bloukanov** [reply]
Thanks for the follow-up! Sure, if it is not too much trouble to release an alpha that would be awesome. We are currently working off of `0.2.2-alpha.3`, I believe for the password decoding as well as support for `Polars >= .8`. Thanks again, this is great news for us.

**Reply by wangxiaoying** [reply]
Sure, the new alpha version has just released: `pip install -U connectorx==0.2.2-alpha.4`. Please feel free to give a try and let me know if you found any issue.

**Reply by bloukanov** [reply]
Thanks for releasing and sharing!

So it appears that this is a bit slower, as of now, for our particular query with multiple CTEs in it, than `pd.read_sql`. Also, interestingly applying a partition slows it down further. Here are some pseudo-test results:

ConnectorX to Polars, no partition: 3m 15s
ConnectorX to pandas, no partition: 3m 17s
ConnectorX to Polars, `row_number` partition in final `select`: I terminated at 5m 30s
ConnectorX to Polars, using built-in binary column as 2-way partition: I terminated at 3m 30s

`pd.read_sql`: 1m 12s

**Reply by wangxiaoying** [reply]
It is possible that the query is slower in ConnectorX on some complex queries since we need to fetch some metadata first, including:
1. min, max query to get range of partition column (if partition)
2. count query to get number of rows in the final result
3. limit 1 query to get the column types of the query result

If the query is simple, these queries usually does not take long. But sometimes if the query is very complex especially when there are nested queries, the database might need to really execute the entire query / subquery and then get the result. It can make these queries as slow as executing the real query.

Can you print out the log here? You can set the `RUST_LOG` environment variable before importing the library like the following example:
```Python
import os
os.environ["RUST_LOG"]="connectorx=debug,connectorx_python=debug"
import connectorx as cx

df = cx.read_sql(mssql_url, query, return_type="polars")
```
This may help us to know which part slows down the most.

**Reply by bloukanov** [reply]
Great, where do I actually find the log..? I set the environment variable prior to import and I can see it when I do `os.environ`:

![image](https://user-images.githubusercontent.com/62404463/141846307-88300dfb-5ca1-4679-9dc6-d5913b01e86c.png)

**Reply by wangxiaoying** [reply]
The log will be printed to the terminal directly.

**Reply by bloukanov** [reply]
Ah, I had to run from the command line. So I am seeing this:

[2021-11-15T20:25:15Z DEBUG connectorx_python::pandas] Running dispatcher
[2021-11-15T20:25:15Z DEBUG connectorx::dispatcher] Fetching metadata
[2021-11-15T20:25:15Z DEBUG connectorx::sql] Transformed limit 1 query:

and then it writes the entire query including all CTEs (I can't paste that here because it's private), placing `select top (1)` in front of the final `select`

[2021-11-15T20:26:20Z DEBUG connectorx::dispatcher] Prepare partitions
[2021-11-15T20:26:41Z DEBUG connectorx::sql] Transformed count query:

again, entire query with all CTEs, and placing `select count * from (final select)`

AS CXTMPTAB_COUNT



[2021-11-15T20:27:47Z DEBUG connectorx::dispatcher] Allocate destination memory
[2021-11-15T20:27:47Z DEBUG connectorx::dispatcher] Create destination partition
[2021-11-15T20:27:47Z DEBUG connectorx::dispatcher] Partition 0, 1008432x3
[2021-11-15T20:27:47Z DEBUG connectorx::dispatcher] Start writing
[2021-11-15T20:28:54Z DEBUG connectorx::dispatcher] Finalize partition 0
[2021-11-15T20:28:54Z DEBUG connectorx::dispatcher] Partition 0 finished
[2021-11-15T20:28:54Z DEBUG connectorx::dispatcher] Writing finished

It appears that it may indeed be executing the entire query 3 times?

**Reply by wangxiaoying** [reply]
Thanks for the info!

> It appears that it may indeed be executing the entire query 3 times?

Yes, it seems like the `top 1` and `count` query basically takes the same time with the real query execution (from Start writing to Writing finished). And I think the min, max query may also takes similar time, which might be the reason that cause your partition becomes slower than not partitioning.

For the count query, we actually just update the code (PR: https://github.com/sfu-db/connector-x/pull/147 -  It is not included in the alpha version yet.) to remove it when the return dataframe is `arrow` (`polars`). And if you need `pandas` dataframe, in this case you can get `arrow` through ConnectorX and convert it into pandas using [to_pandas](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.to_pandas).

For the top 1 query, in postgres and mysql we able to avoid running it by using prepared statement. I'm going to try to do the same thing on mssql.

If partition is on, one way to not running the min max query is to input the `partition_range` column according to the prior knowledge on the data. This range does not necessarily to be accurate since ConnectorX will make the first and last partitioned query open range.

I will release a new alpha version including the above modifications on the count query and top 1 query. Hopefully this new version will improve the performance in your case. Will let you know when it is ready!

**Reply by wangxiaoying** [reply]
Hi @bloukanov , `0.2.2-alpha.5` is released. Please feel free to give it a try and see whether the performance improved.

**Reply by bloukanov** [reply]
Thanks! I will try this out soon and let you know.

**Reply by bloukanov** [reply]
Hi @wangxiaoying, been busy with other stuff  but here are my testing results with `0.2.2-alpha.5`. There is definitely improvement! But still takes a bit longer than `pandas`: 

`connectorx` to `polars`, no partition: 2m 26s
`connectorx` to `polars`, `partition_num` 4, `partition_range` approximate: 3m 15s

`pd.read_sql()`: 1m 20s

**Reply by wangxiaoying** [reply]
Thanks @bloukanov for the test!

I'm suspecting that the improvement is only from the omitted count query. But getting the metadata is still costing the same time with running the entire query. Also I'm not sure why specifying the `partition_range` still cannot make partition faster than no partition. It that's ok, could you please print the log for both no partition and partition with `partition_range` again? Thanks in advance!

**Reply by bloukanov** [reply]
Yup no problem I have them here:

no partition:

```
[2021-11-17T23:16:36Z DEBUG connectorx_python::arrow] Running dispatcher
[2021-11-17T23:16:36Z DEBUG connectorx::dispatcher] Fetching metadata
[2021-11-17T23:17:49Z DEBUG connectorx::dispatcher] Do not need counts in advance
[2021-11-17T23:17:49Z DEBUG connectorx::dispatcher] Allocate destination memory: 0x3
[2021-11-17T23:17:49Z DEBUG connectorx::dispatcher] Create destination partition
[2021-11-17T23:17:49Z DEBUG connectorx::dispatcher] Start writing
[2021-11-17T23:19:02Z DEBUG connectorx::dispatcher] Finalize partition 0
[2021-11-17T23:19:02Z DEBUG connectorx::dispatcher] Partition 0 finished
[2021-11-17T23:19:02Z DEBUG connectorx::dispatcher] Writing finished
```

partition (it seems the whole query is executing for each partition prior to writing?):
```[2021-11-17T23:26:51Z DEBUG connectorx::sql] Transformed single column partition query:``` - now whole query with CTEs, then:
```AS CXTMPTAB_PART WHERE 1 <= CXTMPTAB_PART.pid AND CXTMPTAB_PART.pid < 250001``` and so on for each of the 4 partitions. Then:
```
[2021-11-17T23:26:51Z DEBUG connectorx_python::arrow] Running dispatcher
[2021-11-17T23:26:51Z DEBUG connectorx::dispatcher] Fetching metadata
[2021-11-17T23:27:59Z DEBUG connectorx::dispatcher] Do not need counts in advance
[2021-11-18T15:02:21Z DEBUG connectorx::dispatcher] Allocate destination memory: 0x4
[2021-11-18T15:02:21Z DEBUG connectorx::dispatcher] Create destination partition
[2021-11-18T15:02:22Z DEBUG connectorx::dispatcher] Start writing
[2021-11-18T15:03:52Z DEBUG connectorx::dispatcher] Finalize partition 3
[2021-11-18T15:03:52Z DEBUG connectorx::dispatcher] Partition 3 finished
[2021-11-18T15:04:08Z DEBUG connectorx::dispatcher] Finalize partition 0
[2021-11-18T15:04:08Z DEBUG connectorx::dispatcher] Partition 0 finished
[2021-11-18T15:04:26Z DEBUG connectorx::dispatcher] Finalize partition 1
[2021-11-18T15:04:26Z DEBUG connectorx::dispatcher] Partition 1 finished
[2021-11-18T15:04:26Z DEBUG connectorx::dispatcher] Finalize partition 2
[2021-11-18T15:04:26Z DEBUG connectorx::dispatcher] Partition 2 finished
[2021-11-18T15:04:26Z DEBUG connectorx::dispatcher] Writing finished
```

**Reply by wangxiaoying** [reply]
Thanks for the log! Seems like fetching metadata still needs to execute the query, also the partitioned query takes longer than no partitioning. I suspect it might because all partitioned queries need to execute the same costly subplan first which takes longer time than just execute it once. We will take a look at how to improve it.

**Reply by bloukanov** [reply]
sounds good!
